vowpalwabbit/object_pool.h

```diff
@@ -0,0 +1,98 @@
+#pragma once
+
+#include <set>
+#include <queue>
+
+namespace VW
+{
+
+// Not thread safe
+template <typename T, typename TInitializer>
+struct object_pool
+{
+  object_pool() = default;
+  object_pool(size_t initial_chunk_size, TInitializer initializer = {}, size_t chunk_size = 8)
+    : m_initializer(initializer),
+    m_initial_chunk_size(initial_chunk_size),
+    m_chunk_size(chunk_size)
+  {
+    new_chunk(initial_chunk_size);
+  }
+
+  void return_object(T* obj)
+  {
+    assert(is_from_pool(obj));
+    m_pool.push(obj);
+  }
+
+  T* get_object()
+  {
+    if (m_pool.empty())
+    {
+      new_chunk(m_chunk_size);
+    }
+
+    auto obj = m_pool.front();
+    m_pool.pop();
+    return obj;
+  }
+
+  bool empty() const { return m_pool.empty(); }
+
+  size_t available() const { return m_pool.size(); }
```

Comment:
Why have this function?   It seems conceptually unnecessary?

Reply:
You're right, it isn't really needed anymore. I had it around when it was fixed size